Minimal presto-pinot driver exposing just pinot-common and some helper

agrawaldevesh opened the pull request:
This is so that the accompanying presto-pinot connector in
prestodb/presto proper has simpler dependencies.

This is in line with the approach taken by other heavy weight connectors
like presto-kudu and presto-cassandra.

The basic approach is that the stuff in pinot-common is exposed out as
is and the rest are shaded, since this is the public api. Also added a
class taken from the Pinot broker to federate a PQL to bunch of servers.

The helper class PinotScatterGatherQueryClient does the job of the pinot
to send the query to a pinot server via the internal thrift protocol.

TODO: Couldn't figure out what to do with the logging classes. Currently
they aren't exposed out but what happens if this library need to do some
logging ?

Commit:
Minimal presto-pinot driver exposing just pinot-common and some helper

This is so that the accompanying presto-pinot connector in
prestodb/presto proper has simpler dependencies.

This is in line with the approach taken by other heavy weight connectors
like presto-kudu and presto-cassandra.

The basic approach is that the stuff in pinot-common is exposed out as
is and the rest are shaded, since this is the public api. Also added a
class taken from the Pinot broker to federate a PQL to bunch of servers.

The helper class PinotScatterGatherQueryClient does the job of the pinot
to send the query to a pinot server via the internal thrift protocol.

TODO: Couldn't figure out what to do with the logging classes. Currently
they aren't exposed out but what happens if this library need to do some
logging ?

wenleix commented on pom.xml:
```diff
@@ -0,0 +1,456 @@
+    <modelVersion>4.0.0</modelVersion>
+
+    <parent>
+        <groupId>io.airlift</groupId>
```
Use `com.facebook.airlift` and use version 95 😃

highker commented on src/main/java/com/facebook/presto/pinot/PinotScatterGatherQueryClient.java:
```diff
@@ -0,0 +1,397 @@
+    private static final boolean DEFAULT_EMIT_TABLE_LEVEL_METRICS = true;
+
+    private final String prestoHostId;
+    private final MetricsRegistry registry;
```
This can be a local variable inside the constructor.

highker commented on src/main/java/com/facebook/presto/pinot/PinotScatterGatherQueryClient.java:
```diff
@@ -0,0 +1,397 @@
+    // Connection Pool Related
+    private KeyedPool<PooledNettyClientResourceManager.PooledClientConnection> connPool;
+    private ScheduledThreadPoolExecutor poolTimeoutExecutor;
+    private ExecutorService requestSenderPool;
```
The above 5 variables can be local variables inside the constructor.

highker commented on src/main/java/com/facebook/presto/pinot/PinotScatterGatherQueryClient.java:
```diff
@@ -0,0 +1,397 @@
+        brokerMetrics.initializeGlobalMeters();
+
+        eventLoopGroup = new NioEventLoopGroup();
+        /**
```
Let's make this a comment than a javadoc

highker commented on src/main/java/com/facebook/presto/pinot/PinotScatterGatherQueryClient.java:
```diff
@@ -0,0 +1,397 @@
+        scatterGatherer = new ScatterGatherImpl(connPool, requestSenderPool);
+    }
+
+    public static <T> T doWithRetries(int numRetries, Function<Integer, T> caller)
```
nit: numRetries -> retries

highker commented on src/main/java/com/facebook/presto/pinot/PinotScatterGatherQueryClient.java:
```diff
@@ -0,0 +1,397 @@
+            throw new PinotException(ErrorCode.PINOT_INVALID_PQL_GENERATED, format("Parsing error with on %s, Error = %s", serverHost, e.getMessage()), e);
+        }
+
+        Map<String, List<String>> routingTable = new HashMap<>();
```
Use guava's ImmutableMap.builder

agrawaldevesh replied:
I don't want to bring in a dependency on guava here because then I would have to shade/exclude that etc. Instead I will make the caller (which is just one place in the presto-pinot code) wrap a ImmutableMap.copyOf() around this call.

highker replied:
make sense

highker commented on src/main/java/com/facebook/presto/pinot/PinotScatterGatherQueryClient.java:
```diff
@@ -0,0 +1,397 @@
+            ScatterGatherStats scatterGatherStats = new ScatterGatherStats();
+            CompositeFuture<byte[]> compositeFuture = routeScatterGather(scatterRequest, scatterGatherStats);
+
+            if (compositeFuture == null) {
```
compositeFuture doesn't seems to be nullable

highker commented on src/main/java/com/facebook/presto/pinot/PinotScatterGatherQueryClient.java:
```diff
@@ -0,0 +1,397 @@
+    }
+
+    private Map<ServerInstance, byte[]> gatherServerResponses(
+            String pql,
```
this is not used

highker commented on src/main/java/com/facebook/presto/pinot/PinotScatterGatherQueryClient.java:
```diff
@@ -0,0 +1,397 @@
+            boolean ignoreEmptyResponses,
+            Map<String, List<String>> routingTable,
+            CompositeFuture<byte[]> compositeFuture,
+            ScatterGatherStats scatterGatherStats, boolean isOfflineTable,
```
isOfflineTable is always true; remove?

highker commented on src/main/java/com/facebook/presto/pinot/PinotScatterGatherQueryClient.java:
```diff
@@ -0,0 +1,397 @@
+        @Override
+        public byte[] getRequestForService(List<String> segments)
+        {
+            InstanceRequest r = new InstanceRequest();
```
nit: r -> request

highker replied:
Could you address this comment as well?

highker commented on src/main/java/com/facebook/presto/pinot/PinotScatterGatherQueryClient.java:
```diff
@@ -0,0 +1,397 @@
+     * @return dataTableMap map from server to data table.
+     */
+    private Map<ServerInstance, DataTable> deserializeServerResponses(
+            String pql,
```
not used

highker commented on src/main/java/com/facebook/presto/pinot/PinotScatterGatherQueryClient.java:
```diff
@@ -0,0 +1,397 @@
+     */
+    private Map<ServerInstance, DataTable> deserializeServerResponses(
+            String pql,
+            Map<ServerInstance, byte[]> responseMap, boolean isOfflineTable,
```
isOfflineTable is always true. Also, one param per line, so wrap isOfflineTable in this case

highker commented on src/main/java/com/facebook/presto/pinot/PinotScatterGatherQueryClient.java:
```diff
@@ -0,0 +1,397 @@
+        }
+    }
+
+    private static class ScatterGatherRequestImpl
```
nit: ScatterGatherRequestWrapper

highker replied:
Could you address this comment as well?

highker commented on src/main/java/com/facebook/presto/pinot/PinotScatterGatherQueryClient.java:
```diff
@@ -0,0 +1,397 @@
+        return dataTableMap;
+    }
+
+    private CompositeFuture<byte[]> routeScatterGather(ScatterGatherRequestImpl scatterRequest, ScatterGatherStats scatterGatherStats)
```
use interface ScatterGatherRequest as type than ScatterGatherRequestImpl

highker replied:
Could you address this comment as well?

highker commented on src/main/java/com/facebook/presto/pinot/PinotScatterGatherQueryClient.java:
```diff
@@ -0,0 +1,397 @@
+
+        // Unfortunately the retries will all hit the same server because the routing decision has already been made by the pinot broker
+        Map<ServerInstance, byte[]> serverResponseMap = doWithRetries(pinotRetryCount, (requestId) -> {
+            ScatterGatherRequestImpl scatterRequest = new ScatterGatherRequestImpl(brokerRequest, routingTable, requestId, connectionTimeoutInMs, prestoHostId);
```
use interface ScatterGatherRequest as type than ScatterGatherRequestImpl

highker commented on src/main/java/com/facebook/presto/pinot/PinotScatterGatherQueryClient.java:
```diff
@@ -0,0 +1,397 @@
+        }
+
+        Map<String, List<String>> routingTable = new HashMap<>();
+        routingTable.put(serverHost, new ArrayList<>(segments));
```
nit: new ArrayList<> -> ImmutableList.copyOf

highker commented on src/main/java/com/facebook/presto/pinot/PinotScatterGatherQueryClient.java:
```diff
@@ -0,0 +1,397 @@
+
+        requestSenderPool = Executors.newFixedThreadPool(pinotConfig.getThreadPoolSize());
+        poolTimeoutExecutor = new ScheduledThreadPoolExecutor(50);
+        connPool = new KeyedPoolImpl<>(pinotConfig.getMinConnectionsPerServer(), pinotConfig.getMaxConnectionsPerServer(), pinotConfig.getIdleTimeoutMs(),
```
nit: one param per line
```java
new KeyedPoolImpl<>(
    pinotConfig.getMinConnectionsPerServer(),
    pinotConfig.getMaxConnectionsPerServer(),
    ...

highker commented on src/main/java/com/facebook/presto/pinot/PinotScatterGatherQueryClient.java:
```diff
@@ -0,0 +1,397 @@
+        return defaultBrokerId;
+    }
+
+    public Map<ServerInstance, DataTable> queryPinotServerForDataTable(String pql, String serverHost, List<String> segments, long connectionTimeoutInMs, boolean ignoreEmptyResponses, int pinotRetryCount)
```
nit: connectionTimeoutInMs -> connectionTimeoutInMillis

highker commented on src/main/java/com/facebook/presto/pinot/PinotScatterGatherQueryClient.java:
```diff
@@ -0,0 +1,397 @@
+                    });
+                    throw new PinotException(ErrorCode.PINOT_INSUFFICIENT_SERVER_RESPONSE, String.format("%d of %d servers responded with routing table servers: %s, error: %s", serverResponseMap.size(), routingTable.size(), routingTableForLogging, compositeFuture.getError()));
+                }
+                Iterator<Map.Entry<ServerInstance, byte[]>> iterator = serverResponseMap.entrySet().iterator();
```
Use for loop
```java
for (Map.Entry<...> entry : serverResponseMap.entrySet())

highker commented on src/main/java/com/facebook/presto/pinot/PinotScatterGatherQueryClient.java:
```diff
@@ -0,0 +1,397 @@
+            return serverResponseMap;
+        }
+        catch (ExecutionException | InterruptedException e) {
+            Throwable err = e instanceof ExecutionException ? ((ExecutionException) e).getCause() : e;
```
casting ExecutionException is redundant

highker commented on src/main/java/com/facebook/presto/pinot/PinotScatterGatherQueryClient.java:
```diff
@@ -0,0 +1,397 @@
+            Map<ServerInstance, byte[]> serverResponseMap = compositeFuture.get();
+            if (!ignoreEmptyResponses) {
+                if (serverResponseMap.size() != routingTable.size()) {
+                    Map<String, String> routingTableForLogging = new HashMap<>();
```
Use ImmutableMap.builder

highker commented on src/main/java/com/facebook/presto/pinot/PinotScatterGatherQueryClient.java:
```diff
@@ -0,0 +1,397 @@
+            Map<ServerInstance, byte[]> responseMap, boolean isOfflineTable,
+            String tableNameWithType)
+    {
+        Map<ServerInstance, DataTable> dataTableMap = new HashMap<>();
```
Use ImmutableMap.builder

agrawaldevesh replied:
Replied above on the rationale to not use Guava here.

highker reviewed (commented):
The shading looks great. Minor comments on coding style

agrawaldevesh commented:
I think I have addressed all of the code review comments now and am looking forward to seeing this approved and a version 0.1.0 of this jar released to mavencentral such that I can unblock https://github.com/prestodb/presto/pull/13504. 

Thank you !

highker commented on src/main/java/com/facebook/presto/pinot/PinotScatterGatherQueryClient.java:
```diff
@@ -0,0 +1,363 @@
+        scatterGatherer = new ScatterGatherImpl(connPool, requestSenderPool);
+    }
+
+    public static <T> T doWithRetries(int retries, Function<Integer, T> caller)
```
private

highker commented on src/main/java/com/facebook/presto/pinot/PinotScatterGatherQueryClient.java:
```diff
@@ -0,0 +1,363 @@
+
+        // Unfortunately the retries will all hit the same server because the routing decision has already been made by the pinot broker
+        Map<ServerInstance, byte[]> serverResponseMap = doWithRetries(pinotRetryCount, (requestId) -> {
+            ScatterGatherRequestImpl scatterRequest = new ScatterGatherRequestImpl(brokerRequest, routingTable, requestId, connectionTimeoutInMillis, prestoHostId);
```
nit: `ScatterGatherRequest scatterRequest = ...`

highker reviewed (approved):
LGTM % nits

highker commented:
@wenleix, could you help doing a release? THANKS!
